kong/plugins/ai-prompt-template/handler.lua

```diff
@@ -0,0 +1,131 @@
+local _M = {}
+
+-- imports
+local kong_meta      = require "kong.meta"
+local templater      = require("kong.plugins.ai-prompt-template.templater"):new()
+local fmt            = string.format
+local parse_url      = require("socket.url").parse
+local byte           = string.byte
+local sub            = string.sub
+local cjson          = require("cjson.safe")
+--
+
+_M.PRIORITY = 773
+_M.VERSION = kong_meta.version
+
+
+local log_entry_keys = {
+  REQUEST_BODY = "ai.payload.original_request",
+}
+
+local function bad_request(msg)
+  kong.log.warn(msg)
+  return kong.response.exit(400, { error = { message = msg } })
+end
+
+local BRACE_START = byte("{")
+local BRACE_END = byte("}")
+local COLON = byte(":")
+local SLASH = byte("/")
+
+---- BORROWED FROM `kong.pdk.vault`
+---
+-- Checks if the passed in reference looks like a reference.
+-- Valid references start with '{template://' and end with '}'.
+--
+-- @local
+-- @function is_reference
+-- @tparam string reference reference to check
+-- @treturn boolean `true` is the passed in reference looks like a reference, otherwise `false`
+local function is_reference(reference)
+  return type(reference)      == "string"
```

Comment:
I suggest to cache `type`, `byte`, `sub` and other Lua built-in functions in the top level of the module

Reply:
I don't understand this one

Reply:
This is a LuaJIT performance thing. The builtin `type` function is technically a global variable, and global var access is slower than local var access. By adding this somewhere near the top of the file, you stash the global var in a local reference so that it does not incur the overhead of accessing a global var each time it is used:

```lua
local type = type
```

It's not strictly required and may be overkill in some scenarios (my opinion), but there is really no downside it. In general we recommend doing this with the `ngx` global (`local ngx = ngx`) and most anywhere where there is a particularly hot code path.


(You've already done this for `string.byte` and `string.sub` by nature of having `local byte = string.byte` (etc) at the top of the file, so there is no action needed on that front).

Reply:
The contribution guideline doc is a great resource for things like this:

https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#writing-performant-code